File: backend/app/models/compliance.py
"""设备型号、版本基线与上线合规模型。"""
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, JSON, String, Text, UniqueConstraint
from sqlalchemy.sql import func

from app.database import Base


class DeviceModelProfile(Base):
    """厂商型号能力模板。"""

    __tablename__ = "device_model_profiles"
    __table_args__ = (
        UniqueConstraint("vendor", "model_pattern", "network_type", name="uq_device_model_profile_scope"),
    )

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(150), nullable=False)
    vendor = Column(String(50), nullable=False, index=True)
    model_pattern = Column(String(120), nullable=False, index=True)
    network_type = Column(String(50), nullable=False, default="general", server_default="general", index=True)
    device_type = Column(String(50))
    default_role = Column(String(50))
    capabilities = Column(JSON, nullable=False, default=dict)
    required_checks = Column(JSON, nullable=False, default=list)
    description = Column(Text)
    priority = Column(Integer, nullable=False, default=100, server_default="100")
    is_active = Column(Boolean, nullable=False, default=True, server_default="true")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "vendor": self.vendor,
            "model_pattern": self.model_pattern,
            "network_type": self.network_type,
            "device_type": self.device_type,
            "default_role": self.default_role,
            "capabilities": self.capabilities or {},
            "required_checks": self.required_checks or [],
            "description": self.description,
            "priority": self.priority,
            "is_active": bool(self.is_active),
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }


class VersionBaseline(Base):
    """设备软件版本和补丁基线。"""

    __tablename__ = "version_baselines"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(150), nullable=False)
    model_profile_id = Column(Integer, ForeignKey("device_model_profiles.id", ondelete="SET NULL"), index=True)
    vendor = Column(String(50), index=True)
    model_pattern = Column(String(120), index=True)
    device_role = Column(String(50), index=True)
    allowed_versions = Column(JSON, nullable=False, default=list)
    minimum_version = Column(String(100))
    required_patches = Column(JSON, nullable=False, default=list)
    forbidden_versions = Column(JSON, nullable=False, default=list)
    recommendation = Column(Text)
    priority = Column(Integer, nullable=False, default=100, server_default="100")
    is_active = Column(Boolean, nullable=False, default=True, server_default="true")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "model_profile_id": self.model_profile_id,
            "vendor": self.vendor,
            "model_pattern": self.model_pattern,
            "device_role": self.device_role,
            "allowed_versions": self.allowed_versions or [],
            "minimum_version": self.minimum_version,
            "required_patches": self.required_patches or [],
            "forbidden_versions": self.forbidden_versions or [],
            "recommendation": self.recommendation,
            "priority": self.priority,
            "is_active": bool(self.is_active),
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }


class DeviceComplianceSnapshot(Base):
    """最近一次设备上线合规评估结果。"""

    __tablename__ = "device_compliance_snapshots"

    id = Column(Integer, primary_key=True, index=True)
    device_id = Column(Integer, ForeignKey("devices.id", ondelete="CASCADE"), nullable=False, unique=True, index=True)
    model_profile_id = Column(Integer, ForeignKey("device_model_profiles.id", ondelete="SET NULL"), index=True)
    version_baseline_id = Column(Integer, ForeignKey("version_baselines.id", ondelete="SET NULL"), index=True)
    overall_status = Column(String(30), nullable=False, default="pending", server_default="pending", index=True)
    score = Column(Integer, nullable=False, default=0, server_default="0")
    observed_vendor = Column(String(50))
    observed_model = Column(String(120))
    observed_version = Column(String(255))
    observed_patches = Column(JSON, nullable=False, default=list)
    checks = Column(JSON, nullable=False, default=list)
    blockers = Column(JSON, nullable=False, default=list)
    evaluated_at = Column(DateTime(timezone=True), nullable=False, server_default=func.now(), index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    def to_dict(self):
        return {
            "id": self.id,
            "device_id": self.device_id,
            "model_profile_id": self.model_profile_id,
            "version_baseline_id": self.version_baseline_id,
            "overall_status": self.overall_status,
            "score": self.score,
            "observed_vendor": self.observed_vendor,
            "observed_model": self.observed_model,
            "observed_version": self.observed_version,
            "observed_patches": self.observed_patches or [],
            "checks": self.checks or [],
            "blockers": self.blockers or [],
            "evaluated_at": self.evaluated_at.isoformat() if self.evaluated_at else None,
        }

File: frontend/src/api/compliance.ts
import request from './request'

export type ComplianceStatus = 'compliant' | 'non_compliant' | 'pending' | 'not_monitored'
export type CheckStatus = 'passed' | 'failed' | 'pending' | 'skipped'

export interface DeviceModelProfile {
  id: number
  name: string
  vendor: string
  model_pattern: string
  network_type: string
  device_type?: string
  default_role?: string
  capabilities: Record<string, boolean>
  required_checks: string[]
  description?: string
  priority: number
  is_active: boolean
}

export interface VersionBaseline {
  id: number
  name: string
  model_profile_id?: number
  vendor?: string
  model_pattern?: string
  device_role?: string
  platform_version?: string
  allowed_releases: string[]
  allowed_versions: string[]
  minimum_version?: string
  required_patches: string[]
  forbidden_versions: string[]
  recommendation?: string
  priority: number
  is_active: boolean
}

export interface ComplianceCheck {
  key: string
  label: string
  status: CheckStatus
  message: string
  evidence?: any
  required: boolean
}

export interface DeviceCompliance {
  device_id: number
  model_profile_id?: number
  version_baseline_id?: number
  overall_status: ComplianceStatus
  score: number
  observed_vendor?: string
  observed_model?: string
  observed_version?: string
  observed_patches: string[]
  checks: ComplianceCheck[]
  blockers: Array<{ key: string; label: string; status: string; message: string }>
  evaluated_at: string
  profile?: DeviceModelProfile
  baseline?: VersionBaseline
  device: {
    id: number
    name: string
    ip_address: string
    vendor?: string
    model?: string
    device_role?: string
    is_monitored: boolean
    datacenter?: { id: number; name: string; code?: string }
  }
}

export const getModelProfiles = async () =>
  await request.get('/compliance/model-profiles') as { total: number; items: DeviceModelProfile[] }

export const createModelProfile = async (data: Partial<DeviceModelProfile>) =>
  await request.post('/compliance/model-profiles', data) as DeviceModelProfile

export const discoverModelProfiles = async () =>
  await request.post('/compliance/model-profiles/discover') as { created: number; skipped: number; items: DeviceModelProfile[] }

export const updateModelProfile = async (id: number, data: Partial<DeviceModelProfile>) =>
  await request.put(`/compliance/model-profiles/${id}`, data) as DeviceModelProfile

export const deleteModelProfile = async (id: number) => {
  await request.delete(`/compliance/model-profiles/${id}`)
}

export const getVersionBaselines = async () =>
  await request.get('/compliance/version-baselines') as { total: number; items: VersionBaseline[] }

export const createVersionBaseline = async (data: Partial<VersionBaseline>) =>
  await request.post('/compliance/version-baselines', data) as VersionBaseline

export const updateVersionBaseline = async (id: number, data: Partial<VersionBaseline>) =>
  await request.put(`/compliance/version-baselines/${id}`, data) as VersionBaseline

export const deleteVersionBaseline = async (id: number) => {
  await request.delete(`/compliance/version-baselines/${id}`)
}

export const getComplianceDevices = async (params?: {
  skip?: number
  limit?: number
  overall_status?: string
  vendor?: string
  datacenter_id?: number
  search?: string
  refresh?: boolean
}) => await request.get('/compliance/devices', { params }) as { total: number; items: DeviceCompliance[] }

export const getComplianceSummary = async () =>
  await request.get('/compliance/summary') as {
    total: number
    evaluated: number
    unevaluated: number
    counts: Record<string, number>
    compliance_rate: number
  }

export const getDeviceCompliance = async (deviceId: number, refresh = false) =>
  await request.get(`/compliance/devices/${deviceId}`, { params: { refresh } }) as DeviceCompliance

export const evaluateCompliance = async (deviceId?: number) =>
  await request.post('/compliance/evaluate', undefined, { params: deviceId ? { device_id: deviceId } : undefined }) as {
    total: number
    counts: Record<string, number>
  }
